### PR TITLE
[mlir][mesh] Add TableGen deffinitions of more collective ops

### PR DESCRIPTION
**Pull request by sogartar**
Add definitions for
broadcast, gather, receive, reduce, scatter, send and shift.

I will add verification and canonicalization in another PR. I wanted to see if the signature of the operations is sensible.

### CODE REVIEW

**Comment by sogartar** [comment]
@yaochengji, could you take a look at this PR?

**Comment by yaochengji** [comment]
> @yaochengji, could you take a look at this PR?

Thanks. But I'm currently stuck with some urgent tasks. I'll review it two days later.

File: mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td [hunk 1]
```diff
@@ -339,6 +339,185 @@ def Mesh_AllToAllOp : Mesh_CollectiveCommunicationOpBase<"all_to_all", [
...
+    The result tensor on non-root devices is undefined.
+    Using it will result in undefined behavior.
+
+    Example:
+    ```mlir
+    mesh.cluster @mesh0(rank = 2, dim_sizes = [2, 2])
+    ...
+    %1 = mesh.gather %0 on @mesh0 mesh_axes = [1]
+      gather_axis = 1 root = [1]
+      : (tensor<2x2xi8>) -> tensor<2x4xi8>
```
Comment (yaochengji) [inline mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:414]: Is the output still `tensor<2x4xi8>` if it's not the root?

File: mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td [hunk 2]
```diff
@@ -339,6 +339,185 @@ def Mesh_AllToAllOp : Mesh_CollectiveCommunicationOpBase<"all_to_all", [
...
+  );
+  let assemblyFormat = [{
+    $input `on` $mesh (`mesh_axes` `=` $mesh_axes^)?
+    `gather_axis` `=` $gather_axis
+    `root` `=` custom<DynamicIndexList>($root_dynamic, $root)
+    attr-dict `:` functional-type(operands, results)
+  }];
+}
+
+def Mesh_ReceiveOp : Mesh_CollectiveCommunicationOpBase<"receive", [
```
Comment (yaochengji) [inline mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:458]: Could you rename it from `receive` to `recv`, I know both mpi and nccl use `recv` for concise.

File: mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td [hunk 3]
```diff
@@ -339,6 +339,185 @@ def Mesh_AllToAllOp : Mesh_CollectiveCommunicationOpBase<"all_to_all", [
   let hasCanonicalizer = 1;
 }
 
+def Mesh_BroadcastOp : Mesh_CollectiveCommunicationOpBase<"broadcast", [
+    AllShapesMatch<["input", "result"]>,
```
Comment (yaochengji) [inline mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:343]: Could we use `AllTypesMatch` here?

File: mlir/lib/Dialect/Mesh/IR/MeshOps.cpp [hunk 4]
```diff
@@ -507,6 +508,43 @@ void AllToAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
   patterns.add<EmptyMeshAxesCanonicalizationPattern<AllToAllOp>>(context);
 }
 
+//===----------------------------------------------------------------------===//
+// mesh.broadcast op
+//===----------------------------------------------------------------------===//
+
+LogicalResult
+BroadcastOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  // TODO
```
Comment (yaochengji) [inline mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:517]: I think we should at least add basic test in the `ops.mlir` file even though the verification functions are left as TODO.

File: mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td (hunk 1 above)
Comment (sogartar) [reply mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:414]: If we are doing SPMD there seems to be no other choice. We can say that the results on the other devices is undefined. I don't really like to introduce undefined behavior, maybe we could say that it should be 0-filled, but then we force the runtime to make good on that promise and there should really be no programs that touch the result other than the `root` device.

**Review by yaochengji — APPROVED** [review]

**Comment by MacDue** [comment]
Just an FYI this has broken several buildbots, see:

- https://lab.llvm.org/buildbot/#/builders/61/builds/52686
- https://lab.llvm.org/buildbot/#/builders/181/builds/26327
- https://lab.llvm.org/buildbot/#/builders/268/builds/3685

**Comment by sogartar** [comment]
> Just an FYI this has broken several buildbots, see:
> 
> * https://lab.llvm.org/buildbot/#/builders/61/builds/52686
> * https://lab.llvm.org/buildbot/#/builders/181/builds/26327
> * https://lab.llvm.org/buildbot/#/builders/268/builds/3685

I will fix it.

**Comment by sogartar** [comment]
@MacDue
I have a PR #74359 that should fix it.

**Pull request merged**